### PR TITLE
fix adoption for steady state hosts

### LINKED ISSUE #271: adoption of master nodes fails
After building a new cluster my master nodes in ironic have status "adopt failed" with the error "Error while attempting to adopt node e0759af0-55f0-4e47-a1c7-baeaccd67f61: Cannot validate image information for node e0759af0-55f0-4e47-a1c7-baeaccd67f61 because one or more parameters are missing from its instance_info and insufficent information is present to boot from a remote volume. Missing are: ['ramdisk', 'kernel', 'image_source']."

### PR DESCRIPTION
**Pull request by dhellmann**
Fixes #271

### CODE REVIEW

**Comment by dhellmann** [comment]
We will need a related change in the cluster-api-provider code to always set the image values. I will make those changes after this lands, so I can update the copy of the operator code vendored into the provider repo as part of that PR.

**Comment by dhellmann** [comment]
Without this change, ironic reports the error:

```
Host adoption failed: Error while attempting to adopt node 2a2051e4-a7b9-430d-90ce-c2b9251c8298: Cannot validate image information for node 2a2051e4-a7b9-430d-90ce-c2b9251c8298 because one or more parameters are missing from its instance_info and insufficent information is present to boot from a remote volume. Missing are: ['ramdisk', 'kernel', 'image_source'].
```

**Comment by dhellmann** [comment]
https://github.com/openshift-metal3/dev-scripts/issues/736 is partially related to this change.

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go [hunk 1]
```diff
@@ -555,37 +558,22 @@ func (host *BareMetalHost) NeedsProvisioning() bool {
...
 	if host.Status.Provisioning.Image.URL != "" {
 		// We have an image provisioned.
 		return true
 	}
 	return false
 }
 
-// WasExternallyProvisioned returns true when we think something else
-// is managing the image running on the host.
-func (host *BareMetalHost) WasExternallyProvisioned() bool {
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: Yeah, this is a tough one. I almost wonder if we want ExternallyProvisionedHost to be a different CRD.

File: pkg/controller/baremetalhost/baremetalhost_controller.go [hunk 2]
```diff
@@ -761,10 +761,19 @@ func (r *ReconcileBareMetalHost) actionManageHostPower(prov provisioner.Provisio
...
 	if err != nil {
 		return
 	}
+	if provResult.ErrorMessage != "" {
+		info.host.Status.Provisioning.State = metal3v1alpha1.StateRegistrationError
+		if info.host.SetErrorMessage(provResult.ErrorMessage) {
+			info.publishEvent("RegistrationError", provResult.ErrorMessage)
+			result.Requeue = true
+		}
+		return result, nil
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: If we've failed to adopt the host, I don't think we want to worry about the power anyway, do we? Doesn't a failure here imply that we have bad credentials?

File: pkg/provisioner/ironic/ironic.go [hunk 3]
```diff
@@ -276,6 +276,51 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged bool) (r
...
+			switch err.(type) {
+			case nil:
+			case gophercloud.ErrDefault409:
+				p.log.Info("could not update host settings in ironic, busy")
+				result.Dirty = true
+				return result, nil
+			default:
+				return result, errors.Wrap(err, "failed to update host settings in ironic")
+			}
+		}
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go]: I thought about refactoring that, but decided being explicit here was better. We always know the Op values and can just build a simple array. Go developers seem to be a lot more casual about repeating things. :shrug:

File: pkg/provisioner/ironic/ironic.go [hunk 4]
```diff
@@ -872,8 +872,14 @@ func (p *ironicProvisioner) Adopt() (result provisioner.Result, err error) {
...
-		err = fmt.Errorf("no ironic node for host")
-		return
+		// The node does not exist, but we were called so the
+		// controller thinks that the node existed at one time. That
+		// likely means data loss from restarting the database, so
+		// pass through the validation process to register the node
+		// again. Pass true to indicate that we need to re-test the
+		// credentials, just in case.
+		p.log.Info("re-registering host")
+		return p.ValidateManagementAccess(true)
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go]: Ah, yeah, you might be right about needing to do this more globally. I'll take a look at that.

Adopting an already provisioned server is on my list to do as a separate PR. I think we can just adopt it, though, and don't have to do anything else. Once you've told Ironic a host is adopted then it treats it as though it provisioned it, as far as I can tell. That's why it needs all of the image source details.

File: pkg/controller/baremetalhost/baremetalhost_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zaneb) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: It doesn't imply that AFAIK (currently we're failing because we're not supplying the Image ID, for example, but even once we've fixed that I assume there are other ways it could fail).
We manage the power state on servers that haven't been adopted, so it seems like we should at least attempt to do that when we've failed to adopt.

[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: Looking at the ironic code, the only reasons I can see for a failure are bad credentials or a bug in ironic itself. https://opendev.org/openstack/ironic/src/branch/master/ironic/conductor/manager.py#L2037

[the remark this answers is not in the document]
Comment (juliakreger) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: I concur with @dhellmann. Essentially adoption will require the image fields in the event it is asked to rebuild a host... and otherwise only bmc connectivity is checked. I mean, you could have a transitory RPC failure or something deep under the hood, but trying to manage power state or change it is ill advised and may still fail.

Also, existing power state will not be known until adoption is complete, so it seems kind of odd to try and perform a power action upon failure.

File: pkg/provisioner/ironic/ironic.go [hunk 5]
```diff
@@ -726,7 +771,7 @@ func (p *ironicProvisioner) getUpdateOptsForNode(ironicNode *nodes.Node, checksu
 	if _, ok := ironicNode.InstanceInfo["root_gb"]; !ok {
 		op = nodes.AddOp
 		p.log.Info("adding root_gb")
-	} else if ironicNode.InstanceInfo["root_gb"] != 10 {
+	} else {
```
Comment (juliakreger) [inline pkg/provisioner/ironic/ironic.go]: I'm surprised this is even being set since we're working with whole disk images, root_gb is irrelevant. :\

Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go]: I don't know? The advice I had when originally working on the provisioning logic was that we had to have that value set to something. If we don't need it, we can deal with that in a separate PR.

**Comment by dhellmann** [comment]
There's a related change in https://github.com/metal3-io/cluster-api-provider-baremetal/pull/102

File: pkg/controller/baremetalhost/baremetalhost_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zaneb) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: What I'm hearing is that we shouldn't try to manage the power state of an externally-provisioned host that hasn't been adopted at all?

**Comment by zaneb** [comment]
Docs still [state](https://github.com/metal3-io/baremetal-operator/blob/master/docs/baremetalhost-states.md#externally-provisioned) that the image overrides the externallyProvisioned flag.

File: pkg/controller/baremetalhost/baremetalhost_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: I think that's right, and I think that's what will happen now.

**Comment by dhellmann** [comment]
> > Docs still [state](https://github.com/metal3-io/baremetal-operator/blob/master/docs/baremetalhost-states.md#externally-provisioned) that the image overrides the externallyProvisioned flag.
> 
> Fixed in the latest commit.

Oops, you linked to a different spot than the one I found. Now it's really fixed. :-)

File: pkg/controller/baremetalhost/baremetalhost_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliakreger) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: @zaneb That is correct.

File: pkg/provisioner/ironic/ironic.go (hunk 5 above)
Comment (juliakreger) [reply pkg/provisioner/ironic/ironic.go]: There are a couple different variables, but yeah.. :\

**Comment by dhellmann** [comment]
I added a commit to the end of this PR that handles Ready and Provisioned hosts in the same way.

File: pkg/provisioner/ironic/ironic.go (hunk 5 above)
Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go]: Let's address this question separately.

File: pkg/controller/baremetalhost/baremetalhost_controller.go [hunk 6]
```diff
@@ -317,11 +317,11 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (result re
 	case metal3v1alpha1.StateDeprovisioning:
 		result, err = r.actionDeprovisioning(prov, info)
 	case metal3v1alpha1.StateProvisioned:
-		result, err = r.actionManageHostPower(prov, info)
+		result, err = r.actionManageSteadyState(prov, info)
 	case metal3v1alpha1.StateReady:
-		result, err = r.actionManageHostPower(prov, info)
+		result, err = r.actionManageSteadyState(prov, info)
```
Comment (zaneb) [inline pkg/controller/baremetalhost/baremetalhost_controller.go]: Ready hosts aren't provisioned, if we adopt them in Ironic then Ironic will think they are.

File: pkg/apis/metal3/v1alpha1/baremetalhost_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zaneb) [reply pkg/apis/metal3/v1alpha1/baremetalhost_types.go]: I think a different CRD only works if you assume that the host is running CoreOS and thus never needs to be reprovisioned (it just upgrades itself in-place). In the general case people are likely to want to move an externally-provisioned server under the operator's provisioning control at a later time. That's the source of our dilemma.
This should fix the immediate issue though, and we can continue to iterate on the interface later.

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zaneb) [reply pkg/provisioner/ironic/ironic.go]: Now that we're using this method to also recover from the ironic container being rescheduled, not calling `getUpdateOptsForNode()` means that Ironic will have less information after a resschedule than it did when the host was initially provisioned. Presumably this is not a problem, but I just want to leave a note because it could theoretically be a cause of inconsistencies in Ironic behaviour before and after a reschedule.

File: docs/baremetalhost-states.md [hunk 7]
```diff
@@ -20,8 +20,8 @@ for registration.
 
 An Externally Provisioned host was deployed using another tool and
 then a host object was created with a link to an existing Machine
```
Comment (zaneb) [inline docs/baremetalhost-states.md]: Not part of this change, but we don't require a link to a Machine object any more either.

Comment (dhellmann) [reply docs/baremetalhost-states.md]: Fixed.

File: pkg/controller/baremetalhost/baremetalhost_controller.go (hunk 6 above)
Comment (dhellmann) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: Fixed.

File: pkg/controller/baremetalhost/baremetalhost_controller.go [hunk 8]
```diff
@@ -789,6 +788,39 @@ func (r *ReconcileBareMetalHost) actionManageSteadyState(prov provisioner.Provis
...
+// registered with the provisioner once, so we use
+// ValidateManagementAccess() to ensure that is still true. We don't
+// use Adopt() because we don't want Ironic to treat the host as
+// having been provisioned. Then we monitor its power status.
+func (r *ReconcileBareMetalHost) actionManageReady(prov provisioner.Provisioner, info *reconcileInfo) (result reconcile.Result, err error) {
+
+	// We always pass false for credentialsChanged because if they had
+	// changed we would have ended up in actionRegister() instead of
+	// here.
+	provResult, err := prov.ValidateManagementAccess(false)
```
Comment (zaneb) [inline pkg/controller/baremetalhost/baremetalhost_controller.go]: Question: will we eventually want to do this in the provisioning/deprovisioning state also to handle an Ironic reschedule in that state?

Comment (dhellmann) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: We will need to handle that case, but the way we handle it will be different. If we find that we've re-discovered a node during the middle of a provisioning operation, for example, we will need to restart that operation. The agent relies on the conductor to tell it what to do, so provisioning is going to be broken at that point. So we'll need to power off the host, and start provisioning from scratch. That's going to require some more extensive changes (ValidationManagementAccess will need to tell us somehow that it had to register the host, for example). Since they're more complex, I want to reason through those changes separately.

Comment (zaneb) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: Yeah, not suggesting we tackle it as part of this.
The provisioner's source of truth is Ironic, so if it sees that provisioning hasn't started yet according to the new Ironic it should just start again from scratch. I'd have assumed that Ironic will effect whatever power transitions are necessary to get a host in an unknown initial state to start provisioning, except in the case of adopt.

Comment (dhellmann) [reply pkg/controller/baremetalhost/baremetalhost_controller.go]: The provisioner within the operator will start over, but IPA will be in the middle of an operation and fail the next time it asks for more instructions. I'm not sure if ironic will correctly deduce that the host needs to be restarted if it can talk to an agent there. We'll have to test it and see what happens.
